From 5c49e25deb4bfb2091111c8ee98b978237bc3689 Mon Sep 17 00:00:00 2001
From: Andy Whitcroft <apw@canonical.com>
Date: Wed, 27 Jan 2016 16:05:32 +0000
Subject: [PATCH 0361/1767] UBUNTU: SAUCE: dm: introduce a target_ioctl op to
 allow target specific ioctls

In e56f81e0b01e "dm: refactor ioctl handling" the target specific ioctl
operation was removed in favour of providing a mapping to the underlying
device, to which ioctls are all assumed to apply.  This prevents targets
from having target specific ioctls.

Introduce a new target_ioctl callback which (if present) is offered the
command and arguments for processing.  This callback can return -ENOTTY
to indicate the ioctl should be passed on to the underlying device as
normal.

BugLink: http://bugs.launchpad.net/bugs/1538618
Signed-off-by: Andy Whitcroft <apw@canonical.com>
---
 drivers/md/dm.c               | 9 +++++++++
 include/linux/device-mapper.h | 3 +++
 2 files changed, 12 insertions(+)

diff --git a/drivers/md/dm.c b/drivers/md/dm.c
index 5df4048..9014d46 100644
--- a/drivers/md/dm.c
+++ b/drivers/md/dm.c
@@ -610,6 +610,15 @@ static int dm_blk_ioctl(struct block_device *bdev, fmode_t mode,
 	if (r < 0)
 		return r;
 
+	if (tgt->type->target_ioctl) {
+		int res = tgt->type->target_ioctl(tgt, cmd, arg);
+
+		if (res != -ENOTTY) {
+			r = res;
+			goto out;
+		}
+	}
+
 	if (r > 0) {
 		/*
 		 * Target determined this ioctl is being issued against
diff --git a/include/linux/device-mapper.h b/include/linux/device-mapper.h
index ec1c61c..d770bf0 100644
--- a/include/linux/device-mapper.h
+++ b/include/linux/device-mapper.h
@@ -81,6 +81,8 @@ typedef int (*dm_message_fn) (struct dm_target *ti, unsigned argc, char **argv);
 
 typedef int (*dm_prepare_ioctl_fn) (struct dm_target *ti,
 			    struct block_device **bdev, fmode_t *mode);
+typedef int (*dm_target_ioctl_fn) (struct dm_target *ti, unsigned int cmd,
+				   unsigned long arg);
 
 /*
  * These iteration functions are typically used to check (and combine)
@@ -157,6 +159,7 @@ struct target_type {
 	dm_status_fn status;
 	dm_message_fn message;
 	dm_prepare_ioctl_fn prepare_ioctl;
+	dm_target_ioctl_fn target_ioctl;
 	dm_busy_fn busy;
 	dm_iterate_devices_fn iterate_devices;
 	dm_io_hints_fn io_hints;
-- 
1.9.3

